scripts/helpers.js

```diff
@@ -2,155 +2,159 @@
 
       result += '<a href="' + link + '" class="' + itemClass + '">' + self.__(prefix + text) + '</a>';
-    });
-  });
+    }
+  }
 
   return result;
 });
 
 hexo.extend.helper.register('header_menu', function(className) {
-  var menu = this.site.data.menu;
-  var result = '';
-  var self = this;
-  var lang = this.page.lang;
-  var isEnglish = lang === 'en';
+  const menu = this.site.data.menu;
+  let result = '';
+  const self = this;
+  const lang = this.page.lang;
+  const isEnglish = lang === 'en';
 
-  _.each(menu, function(path, title) {
+  for (let [title, path] of Object.entries(menu)) {
     if (!isEnglish && ~localizedPath.indexOf(title)) path = lang + path;
 
-    result += '<a href="' + self.url_for(path) + '" class="' + className + '-link">' + self.__('menu.' + title) + '</a>';
-  });
+    result += `<a href="${self.url_for(path)}" class="${className}-link">${self.__('menu.' + title)}</a>`;
+  }
 
   return result;
 });
 
 hexo.extend.helper.register('canonical_url', function(lang) {
-  var path = this.page.canonical_path;
+  let path = this.page.path;
   if (lang && lang !== 'en') path = lang + '/' + path;
 
-  return this.config.url + '/' + path;
+  return full_url_for(path);
 });
 
 hexo.extend.helper.register('url_for_lang', function(path) {
-  var lang = this.page.lang;
-  var url = this.url_for(path);
+  const lang = this.page.lang;
+  let url = this.url_for(path);
 
   if (lang !== 'en' && url[0] === '/') url = '/' + lang + url;
 
   return url;
 });
 
-hexo.extend.helper.register('raw_link', function(path) {
-  return 'https://github.com/hexojs/site/edit/master/source/' + path;
-});
+hexo.extend.helper.register('raw_link', (path) => `https://github.com/hexojs/site/edit/master/source/${path}`);
 
 hexo.extend.helper.register('page_anchor', function(str) {
-  var $ = cheerio.load(str, {decodeEntities: false});
-  var headings = $('h1, h2, h3, h4, h5, h6');
+  const $ = cheerio.load(str, {decodeEntities: false});
+  const headings = $('h1, h2, h3, h4, h5, h6');
 
   if (!headings.length) return str;
 
   headings.each(function() {
-    var id = $(this).attr('id');
+    const id = $(this).attr('id');
 
     $(this)
       .addClass('article-heading')
-      .append('<a class="article-anchor" href="#' + id + '" aria-hidden="true"></a>');
+      .append(`<a class="article-anchor" href="#${id}" aria-hidden="true"></a>`);
   });
 
   return $.html();
 });
 
 hexo.extend.helper.register('lunr_index', function(data) {
-  var index = lunr(function() {
+  const index = lunr(function() {
     this.field('name', {boost: 10});
     this.field('tags', {boost: 50});
     this.field('description');
     this.ref('id');
 
-    _.sortBy(data, 'name').forEach((item, i) => {
-      this.add(_.assign({ id: i }, item));
+    const sortBy = (key) => {
+      return (a, b) => {
+        if (a[key] > b[key]) {
+          return 1;
+        } else if (b[key] > a[key]) {
+          return -1;
+        }
+        return 0;
```

Comment:
if `a[key]` is numeric, wouldn't `return a[key] - b[key];` be sufficient? If it is string, can be `return a[key].localeCompare(b[key]);`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

Reply:
It is from [You-Dont-Need-Lodash-Underscore](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_sortby-and-_orderby).
Maybe this is a universal or safe sort in case `a[key]` or `b[key]` is not numeric.

Reply:
I'm sure data.name will always be a string, so `localeCompare` can be utilized. localeCompare has the advantage of supporting non-ascii characters, which suits hexo website.

`return a[key] - b[key]` works on numeric and string.

Reply:
@curbengh 

Strange, `return a[key].localeCompare(b[key]);` or `return b[key].localeCompare(a[key]);` will break the search, while the current code won't.
Maybe we should just follow the guide provided by `You-Dont-Need-Lodash-Underscore`.

Reply:
Maybe it's numeric after all. Use `return a[key] - b[key]` then.

Reply:
@curbengh 

Just add `console.log` to `sort()`:

```js
.sort((a, b) => {
      let x = 0;
      if (a.name > b.name) {
        x = 1;
      } else if (b.name > a.name) {
        x = -1;
      }
      console.log(`${a.name} , ${b.name} | ${a.name - b.name} | ${a.name.localeCompare(b.name)} | ${x} | ${a.name.localeCompare(b.name) === x}`);
      return a.name.localeCompare(b.name);
    })
```

and the output is something like:

```
Prince , Phantom | NaN | 1 | 1 | true
Overdose , pandollo | NaN | -1 | -1 | true
pandollo , orange | NaN | 1 | 1 | true
Overdose , orange | NaN | 1 | -1 | false
pandollo , Paperbox | NaN | -1 | 1 | false
Paperbox , Only | NaN | 1 | 1 | true
pandollo , Only | NaN | 1 | 1 | true
Overdose , Only | NaN | 1 | 1 | true
orange , Only | NaN | 1 | 1 | true
Polk , Prince | NaN | -1 | -1 | true
Prince , phase | NaN | 1 | -1 | false
Polk , phase | NaN | 1 | -1 | false
Prince , Pln | NaN | 1 | 1 | true
Polk , Pln | NaN | 1 | 1 | true
phase , Pln | NaN | -1 | 1 | false
Prince , PolarBear | NaN | 1 | 1 | true
Polk , PolarBear | NaN | 1 | 1 | true
Pln , PolarBear | NaN | -1 | -1 | true
```

Reply:
We shouldn't use `a.name - b.name` directly because it just not works for string.
And `localeCompare()` has diffrent behavior which will break `the sort()`.